gcovr/gcov_parser.py

```diff
@@ -613,9 +626,11 @@ def _add_coverage_for_function(
         )
         return
 
-    function_cov = coverage.function(name)
-    function_cov.lineno = lineno
-    function_cov.count += calls
+    insert_function_coverage(
+        coverage,
+        FunctionCoverage(name, lineno=lineno, call_count=calls),
+        MergeOptions(ignore_function_lineno=True),
```

Comment:
Do you really want to ignore the line number? As I mentioned in https://github.com/gcovr/gcovr/issues/586#issuecomment-1086693394 I prefer to wait if it can be solved on project level.

Reply:
Yes.

My proposed implementation is backwards- (or rather: bug-) -compatible. It maintains exactly the existing behaviour. That is:

* While parsing a .gcov file, the function is attributed to the last line on which it occurs. Therefore, the ignore-function-lineno option is set in this code snippet. This is the equivalent to the old code, which *overwrites* the lineno field:

  ```python
  function_cov = coverage.function(name)
  function_cov.lineno = lineno
  function_cov.count += calls
  ```

* But when merging .json files, this option is not set and the problematic assert in #586 may be triggered.

(To be very pedantic, there is a slight change in behaviour for the .gcov parsing case: The new code uses `max(left.lineno, right.lineno)` to maintain symmetry, whereas the old code just overwrites with the current `lineno`. The current lineno is not guaranteed to increase monotonically due to the presence of template specialization sections. But I'm not sure if that can actually lead to a different outcome.)

Once this PR is merged, I'll think about whether and how #586 can be addressed. Maybe the solution is to also enable this option when merging FileCoverage (on an opt-in basis with a command line flag), maybe the solution is to add a MergeOption to change the concept of function-identity from `name` to `(name, lineno)`. And maybe this is something that can't be supported at all in gcovr.